fix: black screen after multiple initialization

juliansteenbakker opened the pull request:
Fixes #1378
Fixes #1135
Fixes #493 
Fixes #589
Fixes #605 
Fixes #1002
Fixes #1135

Part of #1091, except hot-reloading, which will be fixed in another PR.

This PR fixes the black screen issue, which is caused when the controller is started multiple times (and thus is running), and a piece of code that stops subsequent controller initalization, leaving it uninitialized, and without error message.

The root cause of multiple controllers being started is most likely due to the start() and stop() methods not being awaited for, causing improper controller states. The initialization of the default controller is changed so that async functions are awaited for. 

Users should check their own implementation if a custom controller is being used. This PR makes it easier to discover this because proper error messages are now being shown. (only if kDebugMode = true, otherwise it defaults to a generic error message)

navaronbracke commented on lib/src/mobile_scanner.dart:
```diff
@@ -270,31 +282,39 @@ class _MobileScannerState extends State<MobileScanner>
+  void initState() {
     super.initState();
+    controller = widget.controller ?? MobileScannerController();
+    initController();
```
```suggestion
    unawaited(initController());
```

navaronbracke commented on lib/src/mobile_scanner.dart:
```diff
@@ -270,31 +282,39 @@ class _MobileScannerState extends State<MobileScanner>
-    }
-    // When this widget is unmounted, reset the scan window.
-    unawaited(controller.updateScanWindow(null));
+    disposeCamera();
```
Do we need to await this method, per your comment in the PR description?
If so, then `super.dispose()` should go first, otherwise you'll hit an assert across the async gap (been there done that :p ) Or we can wrap this in `unawaited`

juliansteenbakker replied:
The problem is that if everything is unawaited, race conditions can occur due to the native part receiving multiple calls without waiting for closure. I've changed `super.dispose()` to be on top of the function, and everything seems to be working good right now.

navaronbracke commented on lib/src/mobile_scanner.dart:
```diff
@@ -270,31 +282,39 @@ class _MobileScannerState extends State<MobileScanner>
+    // Dispose default controller if not provided by user
+    if (widget.controller == null) {
+      await controller.dispose();
+      WidgetsBinding.instance.removeObserver(this);
```
Shouldn't we have `WidgetsBinding.instance.removeObserver(this);` done first in this method? What if either stop or dispose fails, then we leak a listener?

juliansteenbakker replied:
Yeah that does not seem right. I have a feeling it is already leaking if the user provides its own controller.

navaronbracke commented:
@juliansteenbakker One small Github related nit, you'll have to put the fixed issues on separate "Fixes #<number>" lines, or it won't pick them all up.

juliansteenbakker commented on lib/src/objects/scanner_error_widget.dart:
```diff
@@ -0,0 +1,40 @@
+              error.errorCode.message,
+              style: const TextStyle(color: Colors.white),
+            ),
+            if (error.errorDetails != null) Text(
```
Did not know about the If-case syntax, cool!

juliansteenbakker commented on lib/src/mobile_scanner.dart:
```diff
@@ -207,12 +209,9 @@ class _MobileScannerState extends State<MobileScanner>
 
         final MobileScannerException? error = value.error;
-
+        // If the controller encountered, show an error screen, or user provided placeholder
```
Bad comment from older code, i'll remove it

juliansteenbakker commented on lib/src/mobile_scanner.dart:
```diff
@@ -270,31 +282,39 @@ class _MobileScannerState extends State<MobileScanner>
   void dispose() {
-    super.dispose();
-
     if (_subscription != null) {
```
Since it's an async function, i think it's better to move it to the other async functions in dispose. Will update the function.

juliansteenbakker commented:
> @juliansteenbakker One small Github related nit, you'll have to put the fixed issues on separate "Fixes #" lines, or it won't pick them all up.

thanks for the tip!

I've added the requested changes, thanks for verifying!

navaronbracke commented:
@juliansteenbakker Sorry if I explained it wrong, but each issue number needs to have the word "fix"/"fixes" in front of it.
I updated the OP to fix that. Also, for the issue that is partially fixed, I updated the wording so Github doesn't auto-close it yet ("partially fixes" still contains the word "fixes")

navaronbracke commented on lib/src/mobile_scanner.dart:
```diff
@@ -270,33 +280,37 @@ class _MobileScannerState extends State<MobileScanner>
+  @override
+  void dispose() {
+    super.dispose();
+    disposeMobileScanner();
```
Should we use unawaited here? (like in initState for the setup?)

```suggestion
    unawaited(disposeMobileScanner());
```

navaronbracke commented on lib/src/mobile_scanner.dart:
```diff
@@ -270,33 +280,37 @@ class _MobileScannerState extends State<MobileScanner>
-      _subscription = null;
-    }
+  Future<void> disposeMobileScanner() async {
+    await _subscription?.cancel();
```
I would move the WidgetsBinding observer call to the font? Also, null-out the subscription?

```suggestion
  Future<void> disposeMobileScanner() async {
    if (widget.controller == null) {
      WidgetsBinding.instance.removeObserver(this);
    }
  
    await _subscription?.cancel();
    _subscription = null;
    
    if (controller.autoStart) {
      await controller.stop();
    }
    
    // Dispose default controller if not provided by user
    if (widget.controller == null) {
      await controller.dispose();
    }    
  }
```

juliansteenbakker replied:
I didn't null it out, because in the `didChangeAppLifecycleState` function, `case AppLifecycleState.inactive:` also doesn't null it out because of syncronous commands. However, maybe we need to create a separate async function for that aswell.

Ultimately it's not really a good pattern of using async methods in `dispose` and `didChangeAppLifecycleState`, but because we are handling a lot on the native side, there isn't really any better option for now. We can look into it for the new major release, or even the next one after that.

navaronbracke replied:
Yeah, async dispose/initState/didChangeAppLifecycleState is meh. So I'd rather go with wrapping them in `unawaited` then.

juliansteenbakker commented:
> @juliansteenbakker Sorry if I explained it wrong, but each issue number needs to have the word "fix"/"fixes" in front of it. I updated the OP to fix that. Also, for the issue that is partially fixed, I updated the wording so Github doesn't auto-close it yet ("partially fixes" still contains the word "fixes")

Thanks for clarification, i see they are linked now. If everything is alright i'll merge this.

navaronbracke commented:
@juliansteenbakker fyI: we should also get this change onto the dev branch after this is merged.

juliansteenbakker commented:
> @juliansteenbakker fyI: we should also get this change onto the dev branch after this is merged.

Yeah that's gonna be the next thing i wanna do, after #1393 gets merged

juliansteenbakker commented on lib/src/mobile_scanner.dart:
```diff
@@ -270,33 +280,41 @@ class _MobileScannerState extends State<MobileScanner>
-      _subscription!.cancel();
-      _subscription = null;
+  Future<void> disposeMobileScanner() async {
+    if (widget.controller == null) {
```
@navaronbracke i don't think we need to change the check here to the same as where `addObserver` is called on https://github.com/juliansteenbakker/mobile_scanner/blob/7b5b2bc9607090de150128d82fb460ca999e5876/lib/src/mobile_scanner.dart#L275

navaronbracke replied:
I agree that the add/ remove of the observer should happen under the same conditions.
So if `onDetect` is set or if the user didn't provide a controller we add/ remove the internal observer?

juliansteenbakker replied:
Yeah, now i look closer to it, i think it's better to change `widget.onDetect != null` to `widget.controller == null`, because if `widget.controller == null`, `widget.onDetect` shouldn't really be null, because that would leave the barcode scanner in a camera state without any options to catch a result. We could even add an assert, that always a controller or onDetect is provided.

juliansteenbakker replied:
The observer in `mobile_scanner` is simply to pause and resume `onDetect` if provided. It doesn't really matter if a controller is also provided or not, the stream should still be paused and resumed. Only when a user doesn't provide `onDetect`, and relies on the own provided `MobileScannerController` to listen to the stream, the `observer` isn't used.

So in short, i don't think we need a check for `widget.controller == null`

juliansteenbakker replied:
We could also just let the subscription be active, even when AppLifecycleState changes, and just leave `controller.start()` and `controller.stop()`.

navaronbracke replied:
> We could also just let the subscription be active, even when AppLifecycleState changes, and just leave `controller.start()` and `controller.stop()`.

That is fine too.

navaronbracke reviewed (approved):
LGTM, if the WidgetsBindingObserver will be removed correctly like you mentioned.

navaronbracke commented:
That was added to fix https://github.com/juliansteenbakker/mobile_scanner/issues/1210
Since with `onDetect` you have a StreamSubscription without an error handler otherwise, and unhandled errors from a Stream bubble up. So I just added an empty error handler, while allowing users to provide their own.

juliansteenbakker commented:
> That was added to fix #1210 Since with `onDetect` you have a StreamSubscription without an error handler otherwise, and unhandled errors from a Stream bubble up. So I just added an empty error handler, while allowing users to provide their own.

Gotya!
